packages/headless/src/features/product-recommendations/product-recommendations-templates-manager.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  ArrayValue,
+  NumberValue,
+  Schema,
+  SchemaValidationError,
+  Value,
+} from '@coveo/bueno';
+import {CoreEngine} from '../../app/engine';
+import {fieldsReducer as fields} from '../../features/fields/fields-slice';
+import {ProductRecommendation} from '../../product-listing.index';
+import {FieldsSection} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {requiredNonEmptyString} from '../../utils/validate-payload';
+import {registerFieldsToInclude} from '../fields/fields-actions';
+import {ProductRecommendationTemplate} from './product-recommendations-templates';
+
+const productRecommendationTemplateSchema =
+  new Schema<ProductRecommendationTemplate>({
+    content: new Value({required: true}),
+    conditions: new Value({required: true}),
+    priority: new NumberValue({required: false, default: 0, min: 0}),
+    fields: new ArrayValue({
+      required: false,
+      each: requiredNonEmptyString,
+    }),
+  });
+
+export interface ProductRecommendationTemplatesManager<Content = unknown> {
+  /**
+   * Registers any number of product recommendation templates in the manager.
+   * @param templates (...ProductRecommendationTemplate<Content>) A list of templates to register.
+   */
+  registerTemplates(
+    ...newTemplates: ProductRecommendationTemplate<Content>[]
+  ): void;
+  /**
+   * Selects the highest priority template for which the given product recommendation satisfies all conditions.
+   * In the case where satisfied templates have equal priority, the template that was registered first is returned.
+   * @param productRec (ProductRecommendation) The product recommendation for which to select a template.
+   * @returns (Content) The selected template's content, or null if no template's conditions are satisfied.
+   */
+  selectTemplate(productRec: ProductRecommendation): Content | null;
+}
+
+/**
+ * A manager in which product recommendation templates can be registered and selected based on a list of conditions and priority.
+ * @param engine (HeadlessEngine) The `HeadlessEngine` instance of your application.
+ * @returns (ProductRecommendationTemplatesManager<Content, State>) A new product recommendation templates manager.
+ */
+export function buildProductRecommendationTemplatesManager<Content = unknown>(
+  engine: CoreEngine
+): ProductRecommendationTemplatesManager<Content> {
+  if (!loadProductRecommendationTemplatesManagerReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const templates: Required<ProductRecommendationTemplate<Content>>[] = [];
+  const validateTemplates = (
```

Comment:
`validateTemplates` could be extracted to the module scope to reduce the scope/complexity of this function.
Just pass the generic along.